apen/apen.1
.TH APEN L "November 24, 1986" "NU Geology"
.\" Define the string Pn to be the device name of the Apple laser printer
.\"    at your site.
.ds Pn apple
.SH NAME
apen \- vector graphics filter for Apple LaserWriter
.SH SYNOPSIS
.B apen
[ flags ] [
.I file ...
]
.SH DESCRIPTION
.I Apen
takes files of the format
.IR plotfil (5)
as input and produces plots on an Apple LaserWriter, or any other PostScript
printer, using the full resolution of the printer.
If no \fIfile\fR is specified, \fIapen\fR reads the standard input.
.PP
The following switches are recognized:
.TP
.B \-e
Produce an EPS file on the standard output.
Implies \fB-vb\fR and \fB-F\fR.
.TP
.B \-v
Produce a plot with the long axis of the paper oriented vertically.  Normally
plots are made with the \fIx\fR-axis along the long side of the paper.  The
base line of the plot is moved down by the width of the paper (approx.\ 8.5").
.TP
.B \-vb
Same as \fB-v\fR, but the base line of the plot is moved to the bottom of the
page of paper (approx.\ 11").  This makes available the bottom 3" or
so of the sheet for plotting.
.TP
.B \-h
Produce a plot with the long axis of the paper oriented horizontally.
This is the default behavior.
.TP
.B \-r
Rotate the orientation to the alternate (vertical if horizontal and vice-versa).
Sometimes this is useful along with \fB-e\fP to get the proper figure
orientation.
.TP
\fB\-c\fP n, \fB\-#\fP n
Plot
.I n
copies.
.TP
.BI \-f \&fatbase
Set line thickness factor equal to
.IR fatbase ;
the greater the value of
.IR fatbase ,
the thicker the lines will be.
This works best when
.I fatbase
is an even number.
.I Fatbase
must be an integer.
.TP
.IB scale x
Set the x-direction scale factor equal to
.IR scale .
.I Scale
may be preceded by an optional
`\-'
which is ignored.
.TP
.IB scale X
Same as
.IB scale x
option.
.TP
.IB scale y
Set y-direction scale factor equal to
.IR scale .
.I Scale
may be preceded by an optional
`\-'
which is ignored.
.TP
.IB scale Y
Same as
.IB scale y
option.
.TP
.I scale
Set both x-direction and y-direction scale factors equal to
.IR scale .
.I Scale
may be preceded by an optional
`\-'
which is ignored.
.TP
.BI \-L\0 dir
Set the directory where
.IR apen 's
data files are located to
.I dir.
This is usually used for testing purposes.
.TP
.BI \-F
Suppress the plot's footer (giving the user name, current directory, file
name and date of printing).
.TP
.BI \-P\0 printer
The output of this program is intended for 
.I printer.
This determines the printer to which the plot is spooled.
If left unspecified, the printer name is taken from the environment variable 
PLOTTER, (if defined), then from PRINTER (if defined), and otherwise defaults 
to \*(Pn.
.TP
.B -t
Send
.IR apen 's
PostScript output to the standard output, instead of spooling it.
.PP
Note that
.I scale
may be specified as either a floating point value or an integer.
Scale factors operate as multipliers to the original
coordinates of points.
If no scale factors are specified, the file will be plotted with the
correct size.
.SH ENVIRONMENT VARIABLES
.TP
PLOTTER
If defined, this is taken to be the name of the printer to which the
plot output is spooled.  -P (see above) overrides this.
.TP
PRINTER
If defined, this is taken to be the name of the printer to which the
plot output is spooled.  -P and PLOTTER (see above) override this.
.TP
LOGNAME
The user name put on the plot footer and the identification in the file
is taken from this environment variable value.
.SH SEE ALSO
plotdb(1), nplot(3), plotfil(5)
.SH AUTHOR
.nf
John Werner, 1988
George Helffrich, 1988, 1990, 2002, 2010
.fi
.SH DIAGNOSTICS
.LP
.I Apen
complains about missing files and invalid plot commands.
.LP
Invalid characters are drawn as a square blot in the output.
.SH BUGS
Raster data currently cannot be plotted, and is replaced by an outline of the
area to be rasterized.  The format for command-line options is messy, but
is compatible with \fItek\fR and \fIipen\fR.


apen/apen.c
/*  Apen -- Program to take pltfil(5) format and print it on any Postscript
	    printer in Postscript.  Refer to a Postscript manual for
	    implementation details.  */
/* Change history:
	    3/88 - Written by John Werner.
	    6/88 - Implement Hershey and hardware-resident fonts.
		   G. Helffrich
	    7/88 - Make this work with the 'pic' program.
	    4/90 - Add -P option
		   G. Helffrich
	    4/90 - Search PRINTER and PLOTTER environment variables
		   G. Helffrich
	    9/90 - Implement -c and -# options in Postscript
	    3/92 - Implement footer on page
	    3/94 - Add color
	    9/02 - EPS output support
	    10/0 - Null printer name allowed
	    10/10 - Alternate data directory for PostScript preamble and
	           nplot font info.
            10/20 - Expand bounding box by +1 point so all lines included.
*/
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <stdio.h>
#include <time.h>	/* added time.h 2021-08-29 crb */
#include <sys/time.h>
#include <pwd.h>
#include "apen.h"

extern int vertical, embedded, pageno;
extern int pxmax, pymax, xmax, ymax, llx, lly, urx, ury;
extern int forcegray;
extern char *progname, *lib;

char *progname, *lib;
void lprfork();

int main(argc,argv)
    int argc; char **argv;
{
    register char *cptr;
    int standard_out=0, 
       triedplot=0, 
       fatbase=0, 
       copies=0;
    int epsfd;
    char *lpdev, *evar;
    char *getenv();
    struct foot_info footer;

    lib = LIB;
    progname = argv[0]; lpdev = PRINTER;

    /* Check if PRINTER environment variable is about.  
       If so, use its value as a plot file name.  */
    evar = getenv("PLOTTER");
    if (evar != NULL)
       lpdev = evar;
    else {
       evar = getenv("PRINTER");
       if (evar != NULL)
	  lpdev = evar;
    }

    /* Set up default footer information. */
    {  time_t now;
       struct timeval tp;
       struct passwd *pwd = (struct passwd *)getpwuid(getuid());
       footer.fi_footer = 1;
       footer.fi_user = malloc(1+strlen(pwd->pw_name));
       strcpy(footer.fi_user,pwd->pw_name);
       footer.fi_dir = getcwd(malloc(MAXPATHLEN),MAXPATHLEN);
       footer.fi_date = malloc(26);
       time(&now);
       strcpy(footer.fi_date,ctime(&now));
    }


    while(--argc) {
	cptr = *++argv;
	if( is_scale(cptr) )
	    continue;
	if (*cptr == '-') {
	    cptr++;
	    switch(*cptr) {
		case '#':
		case 'c':
		    copies = atoi(++cptr);
		    break;
		case 'f':
		    fatbase = atoi(++cptr);
		    break;
		case 'g':
		    forcegray = 1;
		    break;
		case 'h':
		    vertical = 0;
		    /*
		     * Make the x axis on the long side eof the page.
		     * Make sure the Min's and Max's get set right.
		     */
		    break;
		case 'v':
		    vertical = 1;
		    cptr++;
		    if (*cptr == 'b') vertical += 2;
		    /*
		     * Make the y axis on the long side of the page.
		     * Make sure the Min's and Max's get set right.
		     */
		    break;
		case 'e':
		    /*
		     * EPS output
		     */
		    embedded = 1; standard_out = 1; copies = 0; vertical = 3;
		    footer.fi_footer = 0;
		    break;
		case 'r':
		    /*
		     * rotate to other orientation
		     */
		    vertical = !vertical;
		    break;
		case 't':
		    /*
		     * Send output to stdout; -e implies this
		     */
		    standard_out = 1;
		    break;
		case 'L':
		    /*
		     * Set explicit library directory name
		     */
		    lib = cptr+1;
		    break;
		case 'P':
		    /*
		     * Set explicit printer name
		     */
		    lpdev = cptr+1;
		    break;
		case 'F':
		    /*
		     * Suppress footer
		     */
		     footer.fi_footer = 0;
		     break;
		default:
		    fprintf(stderr, "%s: unknown option %s\n", 
		       progname, cptr);
		    exit(1);
		    break;
	    }
	} else 
	    break;
    }

    if (vertical) {
	pxmax = XSIZE;
	pymax = YSIZE;
    } else {
	pxmax = YSIZE;
	pymax = XSIZE;
    }
    xmax = pxmax; ymax = pymax; /* Set clipping bounds to max plot bounds */
    init_color(); /* Initialize color table */

    if (!standard_out)
	lprfork(lpdev);	/* Open up a pipe to "lpr" */

    if (!embedded) 
       printf("%%!PS-Adobe-3.0\n%%%%Pages: (atend)\n");
    else {
       /* Make temporary file, attach to stdout */
       char *epsnm = mktemp("/tmp/apen.XXXXXX"); 
       epsfd = dup(1);
       if (epsnm == NULL || epsfd == -1 ||
          NULL == freopen(epsnm,"w+", stdout)) {
	  fprintf(stderr, "%s: unable to make temporary output file\n", 
		       progname);
	  exit(1);
       }
       unlink(epsnm);
    }
    fprintf(stdout, "%%%%Creator: apen (%s)\n", getenv("LOGNAME"));
    fprintf(stdout, "%%%%CreationDate: %s", footer.fi_date);
       
    dumpfile(INITFILE);		/* Some Postscript procedures we need */

    /* Handle multiple copies if requested */
    if (copies > 0)
       fprintf(stdout, "%%%%BeginSetup\n/#copies %d def\n%%%%EndSetup\n",
	  copies);

    while (argc--) {
	cptr = *argv++;
	triedplot++;
	if(freopen(cptr,"r", stdin) == NULL) {
	    (void) fflush(stdout);
	    fprintf(stderr,"%s:  cannot open %s\n",
	       progname,cptr);
	} else {
	    footer.fi_filename = cptr;
	    doplot(&footer,fatbase);
	}
    }
    if (!triedplot) {
	footer.fi_filename = "standard input";
	doplot(&footer,fatbase);
    }

    if (!embedded) printf("%%%%Trailer\n%%%%Pages: %d\n",pageno);
    dumpfile(TERMFILE);		/* Finish up output nicely */
    printf("%%%%EOF\n");
    if (embedded) {
       /* Insert magic lines, copy temp file to output */
       char bb[1024];
       int n;
       (void)strcpy(bb, "%!PS-Adobe-3.1 EPSF-3.0\n");
       (void)write(epsfd, bb, strlen(bb));
       snprintf(bb, sizeof(bb), "%%%%BoundingBox: %d %d %d %d\n",
          llx-1,lly-1,urx+1,ury+1);
       (void)write(epsfd, bb, strlen(bb));
       snprintf(bb, sizeof(bb), "%%%%Title: %s\n", footer.fi_filename);
       (void)write(epsfd, bb, strlen(bb));
       rewind(stdout);
       for(n = read(fileno(stdout), bb, sizeof(bb));
           n>0;
           n = read(fileno(stdout), bb, sizeof(bb))
       )
          (void)write(epsfd, bb, n);
    }

    return(0);
}


void error(s)
register char *s;
{
   (void) fflush(stdout);
   fprintf(stderr,"%s\n",s);
   exit(1);
}

/* dumpfile()
 * Copy a file into our output stream.  Useful for files that define
 * PostScript procedures and such things.
 */
void dumpfile(s)
char *s;
{
    register FILE *f;
    register int c, d;

    char *fn = malloc(strlen(lib)+strlen(s)+2);
    
    if (!fn) {
        fprintf(stderr, "%s:  Cannot allocate memory for file name.\n",
	    progname);
        exit(1);
    }
    sprintf(fn, "%s%s%s", lib, lib[strlen(lib)] == '/' ? "" : "/", s);

    if ((f = fopen(fn, "r")) == NULL) {
	fprintf(stderr, "%s: couldn't open %s\n", progname, fn);
	exit(1);
    }

    while ((c = getc(f)) != EOF)
	putchar(d=c);

    if (d != '\n') putchar('\n'); (void) fclose(f); (void) fflush(stdout);

    free(fn);
}

int	pipe_des[2];		/* pipe_des[0] is the pipe output,
				   pipe_des[1] is the input */
void lprfork(pname)
   char *pname;
{
	int child;

	/* Open pipe to lpr	 */
	if (pipe(pipe_des) != 0) {
		fprintf(stderr, "Cannot open pipe to lpr.\n");
		exit(1);;
	}

	/* fork off lpr */
	if ((child = fork()) < 0) {
		fprintf(stderr, "Cannot fork off lpr.\n");
		exit(1);
	}

	if (child == 0) {
		char printer[128], *lparg;

		if (strlen(pname)) {
#ifdef BSD
			sprintf(printer,"-P%s",pname);
#else
			sprintf(printer,"-d%s",pname);
#endif
			lparg = printer;
		} else
			lparg = NULL;
		/*
		 * This code block runs as a child process
		 */
		/* Close unused input end of pipe; "lpr" doesn't need it. */
		close(pipe_des[1]);

		/* Attach output end of pipe to the standard input of "lpr" */
		dup2(pipe_des[0], 0);

		/* exec lpr */
#ifdef BSD
		execl(LPR, "lpr", lparg, (char *)0);
#else
		execl(LPR, "lp", lparg, (char *)0);
#endif

		/* if we get here something is wrong */
		fprintf(stderr, "Cannot exec lpr.\n");
		exit(1);
	}

	/* Close unused output end of pipe; "apen" doesn't need it. */
	close(pipe_des[0]);

	/*
	 * Attach input end of pipe to our standard output.
	 * This may not be the most portable thing in the world.
	 */
	dup2(pipe_des[1], fileno(stdout));
}
